bookdown/03-optimization.Rmd

```diff
@@ -231,6 +231,231 @@ Note that we do not expect any differences here compared to the non-tuned approa
+So, $1/\eta$ is the fraction of remaining configurations after a bracket removes his worst performing ones, but $\eta$ is also the factor by that the budget is increased for the next stage.
+Because there is a different maximum budget per configuration that makes sense in different scenarios, the user also has to set this as the $R$ parameter.
+No further parameters are required for Hyperband -- the full required budget across all brackets is indirectly given by $$(\lfloor \log_{\eta}{R} \rfloor + 1)^2 * R$$ [@Li2016].
+To give an idea how a full bracket layout might look like for a specific $R$ and $\eta$, a quick overview is given in the following table.
+
+```{r, eval = TRUE, echo = FALSE}
+eta = 2
+R = 8
+result = data.frame()
+smax = floor(log(R, eta))
+B = (smax + 1) * R
+
+# outer loop - iterate over brackets
+for (s in smax:0) {
+
+  n = ceiling((B/R) * ((eta^s)/(s+1)))
+  r = R * eta^(-s)
+
+  # inner loop - iterate over bracket stages
+  for (i in 0:s) {
+
+    ni = floor(n * eta^(-i))
+    ri = r * eta^i
+    result = rbind(result, c(smax - s + 1, i + 1, ri, ni))
+  }
+}
+
+names(result) = c("bracket", "stage", "budget", "n")
+
+knitr::kable(
+  list(
+	result[1:4, 2:4],
+	data.frame(stage = 1:3, budget = 2^(1:3), n = c(6,3,1)),
+	data.frame(stage = 1:2, budget = c(4,8), n = c(4,2)),
+	data.frame(stage = 1, budget = 8, n = 4)
+  ),
+  booktabs = TRUE,
+  caption = "Hyperband layout for $\\eta = 2$ and $R = 8$, consisting of four brackets with $n$ as the amount of active configurations."
+)
+```
+
+Of course, early termination based on a performance criterion may be disadvantageous if it is done too aggressively in certain scenarios.
+A learner to jumping radically in its estimated performance during the training phase may get the best configurations canceled too early, simply because they do not improve quickly enough compared to others.
+In other words, it is often unclear beforehand if having an high amount of configurations $n$, that gets aggressively discarded early, is better than having a high budget $B$ per configuration.
+The arising tradeoff, that has to be made, is called the "$n$ versus $B/n$ problem".
+To create a balance between exploitation by selection based on early training performance versus exploration of training performances in later training stages, $\lfloor \log_{\eta}{R} \rfloor + 1$ brackets are constructed with an associated set of varyingly sized configurations.
+Thus, some brackets contain more configurations, with a small initial budget.
+In these, a lot are discarded after having been trained for only a short amount of time, corresponding to the exploitative trainer in our horse analogy.
+Others are constructed with fewer configurations, where discarding only takes place after a significant amount of budget was consumed.
+The last bracket usually never discards anything, but also starts with only very few configurations -- this is equivalent to the explorative trainer.
```

Comment:
For me it is a little bit unintuitive that the trainer that starts with few explorations is explorative? He does not "explore" many horses, but he does exploit only a couple of hourses... 
Do I misunderstand that? :-/

Reply:
removed every "exploration vs exploitation" reference in the text.
The trainers are now either "selective" aka "explorative in earlier stages" or "non-selective" aka "explorative of later training stages"